core/src/test/java/org/apache/iceberg/TestSchemaUpdate.java

```diff
@@ -1388,34 +1395,39 @@ public void testSetIdentifierFieldsFails() {
             .setIdentifierFields("locations.key.zip")
             .apply());
 
-    AssertHelpers.assertThrows("add a nested field in list should fail",
```

Comment:
How about this test, it should not need to be removed right?  (if i understand, the first check is still about list).  

Seems the new test "required_list.element.x" is about required list, this can be kept and test about optional list.

Reply:
Yes it still fails, only now the reason it fails is because the hidden middle  field `element` is optional:
> "Cannot add field x as an identifier field: must not be nested in an optional field 14: element: optional struct<15: x: required long, 16: y: required long>"
to contain:

This test is against "list" type, so I added the "required_list.element.x" test to make sure the Precondtion will fail for the "list" type reason, to replace it.

Reply:
Ah i see,  it checks from bottom upward , and hits the "element" optional check first.  What do you think, we still keep the test as "add a nested field in optional list should fail", and change the assert message?  This test  may still have value as there's not another that test this case, unless I am missing seeing it.

Reply:
I think we can keep this test, and I think changing the validation logic might be a better option than changing the assert message. 
I feel we shouldn't expose the `element` or  `key-value` field in Preconditions error message because these fields are invisible to the users and may confuse them. I think a validation from top to bottom can give a more correct error message. 
```
Deque<Integer> deque = Lists.newLinkedList();
while (parentId != null) {
  deque.push(parentId);
  parentId = idToParent.get(parentId);
}

while (!deque.isEmpty()) {
  Types.NestedField parent = idToField.get(deque.pop());
  Preconditions.checkArgument(parent.type().isStructType(),
      "Cannot add field %s as an identifier field: must not be nested in %s", field.name(), parent);
  Preconditions.checkArgument(parent.isRequired(),
      "Cannot add field %s as an identifier field: must not be nested in an optional field %s",
      field.name(), parent);
}
```
WDYT?

Reply:
Yea, that sounds good to me

Reply:
Maybe can add a small code comment, like 'Exploring from root for better error message for list and map types'?

Reply:
Will add.

Reply:
Done.